api/src/main/java/org/apache/iceberg/DeletedRowsScanTask.java

```diff
@@ -0,0 +1,30 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg;
+
+/**
+ * A scan task for deleted data records generated by adding delete files to the table.
+ */
+public interface DeletedRowsScanTask extends ChangelogScanTask {
```

Comment:
I've discovered one issue implementing this API locally. Specifically, we need to differentiate between historical deletes and new deletes in this task. If we add a new delete file, we should NOT output records that were deleted in earlier snapshots. We need to apply old deletes first and then see what was matched with the new deletes. It's not something we can express with `FileScanTask`. I could add another method for new deletes and return deletes from earlier snapshots in the existing `deletes()` method. That sounds awkward.

Reply:
@aokolnychyi is it necessary to suppress the new deletes? We got new deletes because some writer deleted the same row again.

Reply:
Are you talking about position deletes within the same snapshot?

I thought for that we were going to check the sequence number and adjust behavior. So there's no need for a flag.

Reply:
@stevenzwu, not sure yet. I guess that would be the correct behavior but may complicate things quite a bit as we need to load historic deletes too.

Reply:
Think a bit more. We have 2 options.
Option 1. Reader side change, based on what Anton proposed.
    1. Need changes for deletes() in FileScanTask to return a list of historical delete files as well as current snapshot's delete files.
    2. Output only if rows are deleted in the current deletes are not in the previous ones.
    3. Need changes from the readers. Actually it needs a special reader only for this purpose, which return the deleted rows in this snapshot but not in the previous snapshots. This part of logic cannot be reused by `_deleted` metadata column. 

For example, a change like this:
```
  public static <T> CloseableIterable<T> markDeleted(CloseableIterable<T> rows, Predicate<T> isDeleted,
                                                     Consumer<T> deleteMarker) {
    return CloseableIterable.transform(rows, row -> {
      if (isDeleted.test(row) && isNotInHistoricalDelete.test(row)) {
        deleteMarker.accept(row);
      }
      return row;
    });
  }
```

Option 2. Use the Spark dataframe to deduplicate
    1. Read previous deleted rows into a dataframe df1, read the current deleted rows into a datadrame df2
    2. `df2.exceptAll(df1)` to deduplicate
    3. we need a new scan task for previous deletes, maybe just a minor change of the current DeletedRowsScanTask. For example, for data file f1, s1 added delete file d1, and s2 added delete file d2, and s3 added delete file d3, assume s3 is the current snapshot the scan task for previous deletes should has f1, d1, d2.
    4. We need to output pos and data file name to identify the duplication.
    
Option 1 needs changes from reader, but it should performance better than option 2. Option 2 need less code change for scan task and reader, but it may not perform well, and cannot be used across engines.

Reply:
I am not sure we need to de-dup the delete marker. If `delete from table where key=k1` is called twice, that should be fine. 2nd delete would be no-op.

Reply:
Let's say we delete a row r1 in s1, and delete it again in s2. We will get two CDC delete records, one from s1, another from s2, as you pointed out. The second CDC delete record could be considered as a no-op, which is probably fine(I will come back with an edge case). The problem happens when we only generate CDC records from s2. How do we handle the deleted row r1 in s2? Do we output it or not? It is confusing if we output it since it is really a no-op.

The edge case is that there are two same rows in the data files(r1, r2). We don't know whether r1 is deleted twice or both r1 and r2 are deleted if we output duplicated delete cdc records. Unless we project row pos and data file name in cdc records, which sounds weird, and exposes two much internal implementation details.

Reply:
> The problem happens when we only generate CDC records from s2

I feel it is an invalid case. CDC read should been incremental and continuous. If an Iceberg table is the CDC source, I would imagine CDC read should start with full table scan, then transition to the incremental changelog scan. We shouldn't read S2 without reading S1.

> The edge case is that there are two same rows in the data files(r1, r2). We don't know whether r1 is deleted twice or both r1 and r2 are deleted if we output duplicated delete cdc records

I didn't quite get this edge case. Do r1 and r2 have the same primary key and the data file is removed?

Reply:
It has nothing to do with primary key(s). Iceberg allows two or multiple rows to be identical. For example,  here are two rows with the same values in each columns(id, name, age).
```
r1 (1, "Alice", 20) at pos 0 of data file df1
r2 (1, "Alice", 20) at pos 1 of data file df1
```
Assume both s1 and s2 delete r1,  then we will emit two delete records.
```
(1, "Alice", 20, "D", "S1")
(1, "Alice", 20, "D", "S2")
```
The user won't be able to distinguish the following two scenarios. 
1. Both deleted r1.
2. One deleted r1, while another deleted r2.

Reply:
Without applying historical deletes, the changelog may tell the user that a particular snapshot deleted a record that was wasn't live and was removed in an earlier operation. That can happen both with equality and position deletes.

Reply:
> It has nothing to do with primary key(s). Iceberg allows two or multiple rows to be identical. For example, here are two rows with the same values in each columns(id, name, age).
> 
> ```
> r1 (1, "Alice", 20) at pos 0 of data file df1
> r2 (1, "Alice", 20) at pos 1 of data file df1
> ```

For V2 format should not exist two same data but no delete data. It's invalid.

> 
> The user won't be able to distinguish the following two scenarios.
> 
> 1. Both s1 and s2 deleted r1.
> 2. S1 deleted r1, while s2 deleted r2. Or vise-verse.

Base on the above this case is invalid. Even this happen, for equal-delete there should be `Both s1 and s2 deleted r1.`, for postion-delete there is specific data what locate at what file and postion should be delete.

Reply:
I think either interpretation of the concurrent delete case is fine. For the position delete example, d1 and d2 are concurrent and based on the same underlying data, so I'm fine saying BOTH of them deleted a row. It is a little strange to say d1 deleted pos 0 and d2 didn't because d2 actually did encode that delete and would have deleted row 0 if it had won the race to commit. But there is also a strong argument that we should produce a delete for each row once to avoid confusion when people consume these tables.

I think as long as the implementation documents what it does, either one is fine.